### PR TITLE
refactor: migrate to Jira Cloud API v3 with enhanced SAST reporting

### PR DESCRIPTION
**Pull request by AjayJagan**

### CODE REVIEW

**Review by grdryn — COMMENTED** [review]
## AI Code Review

> This review was performed by **Claude Sonnet 4.6** (`claude-sonnet-4-6`). It is provided for informational purposes and does not constitute a human approval or sign-off.

---

## Code Review: `refactor: migrate to Jira Cloud API v3 with enhanced SAST reporting`

**Scope:** 47 files, +3844/-1009 | 3 commits

---

### Overview

Full rewrite from a monolithic script to a proper installable Python package (`src/snyk_jira_reporter/`), migrating from `issues.redhat.com` (Jira Server) to `redhat.atlassian.net` (Jira Cloud API v3), adding SAST vulnerability support, and introducing a CI pipeline with tests.

---

### Architecture & Structure

**Positive changes:**
- Clean subpackage layout: `clients/`, `config/`, `models/`, `services/`, `utils/`, `exceptions/`
- `AppSettings` via `pydantic-settings` replaces scattered `os.getenv()` calls — required fields fail fast with clear errors
- Proper exception hierarchy (`SnykJiraReporterError` → `SnykClientError`, `JiraClientError`, etc.)
- Duplicate repo detection in `load_component_mapping` (previously silent data corruption)
- JSON Schema validation for `jira_components_mapping.json` with CI step

---

### Critical Issues

**1. Unused `JIRA` library instance (`jira_client.py:__init__`)**

`self.client = JIRA(...)` is constructed but **never called** — all Jira operations use raw `requests`. Yet `jira>=3.8.0` remains a production dependency and is instantiated at startup (network roundtrip). Either use it consistently or remove it entirely.

**2. ADF/wiki-markup mismatch (`vulnerability.py:jira_description`)**

The description string uses Jira wiki markup (`*project_name*`, `\n\n` newlines), but `_convert_issue_fields_for_cloud` wraps it in a single flat ADF paragraph node. In Jira Cloud, `*text*` renders as literal asterisks, not bold. Descriptions will look broken. Either build ADF natively in `jira_description()` or strip the markup.

**3. Empty component name sent to Jira (`jira_client.py:create_jira_issues`)**

```python
"components": [{"name": vulnerability.component}],
```

When no mapping exists, `vulnerability.component` is `""`, sending `[{"name": ""}]` — this will cause a 400 error from Jira. Guard it:

```python
if vulnerability.component:
    issue_dict["components"] = [{"name": vulnerability.component}]
```

**4. Unverified API endpoint (`jira_client.py:_search_issues_v3`)**

```python
url = f"{self.jira_server}/rest/api/3/search/jql"
```

`/rest/api/3/search/jql` does not appear in the official Jira Cloud REST API v3 spec — it should likely be `/rest/api/3/search`. This will 404 in production.

**5. Non-existent GitHub Actions versions (`ci.yaml`)**

`actions/checkout@v6` and `actions/setup-python@v6` don't exist (latest are v4 and v5 respectively). These will fail or resolve unpredictably.

---

### Other Issues

| Location | Issue |
|---|---|
| `jira_client.py:_search_issues_v3` | `timeout=30` hardcoded; should use `DEFAULT_REQUEST_TIMEOUT_SECONDS` |
| `jira_client.py:get_existing_jira_for_project` | Off-by-one pagination: loop makes one extra empty request on page-aligned results. Check `len(next_page) < max_results` or use response `total` field |
| `jira_client.py:get_existing_jira_for_project` | `except Exception` swallows `KeyboardInterrupt`/`SystemExit` — narrow to specific types |
| `jira_client.py` & `vulnerability_service.py` | `_extract_text_from_adf` is duplicated — move to `utils/` |
| `constants.py` | `JIRA_SECURITY_FIELD_ID = "10034"` is instance-specific — make configurable via `AppSettings` |
| `settings.py` | `jira_project_key: str = "RHOAIENG"` bakes in a project-specific default — should be required or at least documented |
| `__main__.py` | `--limit` is typed as `str` but must be an integer for the Snyk API — use `type=int` |
| `snyk_client.py:_paginate` | `SNYK_API_BASE_URL + next_link` breaks if Snyk returns an absolute URL — use `urllib.parse.urljoin` |
| `validate_config.py` | Emoji (`❌ ✅ ⚠️`) in CI output — use plain ASCII for reliability |

---

### Test Coverage

**Good:** Snyk client pagination edge cases, `segregate_issues` logic, settings validation, component mapping inversion/duplicate detection.

**Gaps:**
- `_convert_issue_fields_for_cloud` has no tests (critical given the ADF conversion)
- `_transition_issue_v3` and `_add_comment_v3` untested
- `process_projects` orchestration loop (project type routing: sast/dep/dockerfile) untested
- No test covering the empty-component bug described above

---

### Summary

This is a substantial and well-intentioned improvement. The package structure, settings validation, exception hierarchy, and test suite are all real wins. The blocking concerns before merge are items 1–5 above — particularly the ADF rendering bug (descriptions will look broken in Jira Cloud), the empty-component 400 error, and the unverified search endpoint.

**Comment by AjayJagan** [comment]
Hi Ger!, Thanks for the  review! I've addressed the issues now:

I have restored the jira library(python) as it has better capabilities and lesser boilerplate

File: config/jira_components_mapping.json [hunk 1]
```diff
@@ -1,51 +1,95 @@
...
+        "red-hat-data-services/odh-dashboard"
+    ],
+    "AI Core Platform": [
+        "red-hat-data-services/rhods-operator",
+        "red-hat-data-services/kube-auth-proxy"
+    ],
+    "Model Runtimes": [
+        "red-hat-data-services/openvino_model_server",
+        "red-hat-data-services/text-generation-inference",
+        "red-hat-data-services/caikit",
```
Comment (spolti) [inline config/jira_components_mapping.json:46]: isn't caikit depracated? or are you keeping for older branches?

File: config/jira_components_mapping.json [hunk 2]
```diff
@@ -1,51 +1,95 @@
...
-    "red-hat-data-services/spark-operator": "Spark Operator",
-    "opendatahub-io/ai-gateway-payload-processing": "Inference Gateway"
+    "Model Serving": [
+        "red-hat-data-services/kserve",
+        "red-hat-data-services/modelmesh-serving",
+        "red-hat-data-services/odh-model-controller",
+        "red-hat-data-services/modelmesh",
+        "red-hat-data-services/rest-proxy",
+        "red-hat-data-services/modelmesh-runtime-adapter"
+    ],
```
Comment (spolti) [inline config/jira_components_mapping.json:9]: will the llm-d components be added?

Comment (grdryn) [reply config/jira_components_mapping.json:9]: @spolti I'd expect that the changes here in this file are a transformation of the existing file that's already there on main. Are the entries that you're looking for already there on main? I would not expect that adding new or removing entries would be part of this logic rewrite change: adding entries is the responsibility of the owning team, as mentioned in the description of this task in the tech preview template epic: [RHOAIENG-30974](https://redhat.atlassian.net/browse/RHOAIENG-30974), so please create a separate PR(s) if there are discrepancies that you notice.

File: config/jira_components_mapping.json (hunk 1 above)
Comment (AjayJagan) [reply config/jira_components_mapping.json:46]: we could do a clean up, but for now I was just migrating the old bit list to a better list

File: config/jira_components_mapping.json (hunk 2 above)
Comment (AjayJagan) [reply config/jira_components_mapping.json:9]: Agreed with @grdryn . This is just a migration of the old list, for new ones we can always add them to main and until the pr is merged, I will take care of putting them in the right group. After that it would be the owning teams again

File: src/snyk_jira_reporter/clients/jira_client.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+        created_count = 0
+        for vulnerability in vulnerabilities_to_create:
+            try:
+                has_component = bool(vulnerability.component)
+                labels = create_labels(vulnerability, has_component_mapping=has_component)
+
+                # Prepare issue fields for jira library
+                issue_fields = {
+                    "project": {"key": jira_project_id},
```
Comment (grdryn) [inline src/snyk_jira_reporter/clients/jira_client.py]: Is this correct, or should it instead be:


```suggestion
                    "project": {"id": jira_project_id},
```

or:


```suggestion
                    "project": {"key": jira_project_key},
```

File: src/snyk_jira_reporter/clients/jira_client.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+        Returns:
+            ADF formatted description dict.
+        """
+        # Strip wiki markup to prevent rendering issues in Jira Cloud
+        clean_text = self._strip_wiki_markup(text)
+        return {
+            "type": "doc",
+            "version": 1,
+            "content": [{"type": "paragraph", "content": [{"type": "text", "text": clean_text}]}],
```
Comment (grdryn) [inline src/snyk_jira_reporter/clients/jira_client.py]: Depending on whether the input text is a single paragraph or not, coercing it all into a single paragraph node here might affect readability. Not something to change here now, but something to look out for on issues that get created over time.

File: src/snyk_jira_reporter/services/vulnerability_service.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+        old_uid = match.group(1).strip()
+        normalized_old = _normalize_uid_for_comparison(old_uid)
+
+        logger.debug(
+            "  Found existing issue %s with UID: %s (normalized: %s)", jira.get("key"), old_uid, normalized_old
+        )
+
+        # Check if this old issue matches any current vulnerability
+        if normalized_old in new_vuln_normalized:
```
Comment (grdryn) [inline src/snyk_jira_reporter/services/vulnerability_service.py:199]: The new "normalized" strings appear to drop information (snyk id for old format, issue id for new). What if there are multiple vulnerabilities in the same file (either in Snyk, or multiple old issues in jira that need to be cleaned up)? I've got a feeling that in one case we'll only get one issue from the multiple in a file, and for the other case, there will be leftover issues. WDYT?

File: scripts/validate_config.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+"""Validate configuration files against their JSON schemas."""
```
Comment (grdryn) [inline scripts/validate_config.py:2]: Potential future enhancement: also check that the Jira components actually exist, and git repos exist. It's frustratingly common for the jira components especially to be renamed.

File: .github/workflows/jira-snyk.yaml [hunk 7]
```diff
@@ -6,31 +6,37 @@ on:
...
   # secrets
   JIRA_API_TOKEN: ${{ secrets.jira_api_token }}
+  JIRA_EMAIL: ${{ secrets.jira_email }}
   SNYK_API_TOKEN: ${{ secrets.snyk_api_token }}
-  #JIRA envs
-  JIRA_SERVER: https://issues.redhat.com # Org jira server
+  # JIRA envs
+  JIRA_SERVER: https://redhat.atlassian.net # Jira Cloud server
   JIRA_PROJECT_ID: 12340620 # refers to RHOAIENG
+  JIRA_PROJECT_KEY: RHOAIENG
```
Comment (grdryn) [inline .github/workflows/jira-snyk.yaml:14]: Potential future enhancement: support multiple JIRA spaces (e.g. RHOAIENG, RHAIENG, INFERENG, etc). I guess that would probably pull these constants out to the component mapping config somehow

File: scripts/validate_config.py (hunk 6 above)
Comment (AjayJagan) [reply scripts/validate_config.py:2]: Noted: will do it as a separate task(will create a jira)

File: .github/workflows/jira-snyk.yaml (hunk 7 above)
Comment (AjayJagan) [reply .github/workflows/jira-snyk.yaml:14]: Noted: will do it as a separate task(will create a jira)

File: src/snyk_jira_reporter/clients/jira_client.py (hunk 4 above)
Comment (AjayJagan) [reply src/snyk_jira_reporter/clients/jira_client.py]: changed this now

File: src/snyk_jira_reporter/services/vulnerability_service.py (hunk 5 above)
Comment (AjayJagan) [reply src/snyk_jira_reporter/services/vulnerability_service.py:199]: yea that would be a problem. I fixed it now

**Comment by AjayJagan** [comment]
@grdryn  I believe I have addressed your comments. Let me know if it looks good to you

File: README.md [hunk 8]
```diff
@@ -1,20 +1,398 @@
...
+- Python 3.10+
+- Snyk API token with org access
+- Jira Cloud API token with project access
+- Access to RHOAIENG Jira project
+
+### Installation
+
+```bash
+# Install the package
+pip install -e .
```
Comment (spolti) [inline README.md:27]: It worth migrate to `uv` :)

Comment (AjayJagan) [reply README.md:27]: will do

**Review by grdryn — APPROVED** [review]

File: README.md (hunk 8 above)
Comment (grdryn) [reply README.md:27]: Indeed, `uv` is very nice, and would be a nice future enhancement to this 👍
